clippy_lints/src/use_crate_prefix_for_self_imports.rs

```diff
@@ -0,0 +1,149 @@
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// This lint checks for imports from the current crate that do not use the `crate::` prefix.
+    /// It suggests using `crate::` to make it clear that the item is from the same crate.
+    ///
+    /// ### Why is this bad?
+    /// When imports from the current crate lack the `crate::` prefix, it can make the code less readable
+    /// because it’s not immediately clear if the imported item is from the current crate or an external dependency.
+    /// Using `crate::` for self-imports provides a consistent style, making the origin of each import clear.
+    /// This helps reduce confusion and maintain a uniform codebase.
+    ///
+    /// ### Example
+    /// ```rust,ignore
+    /// // lib.rs
+    /// mod foo;
+    /// use foo::bar;
+    /// ```
+    ///
+    /// ```rust,ignore
+    /// // foo.rs
+    /// #[path = "./foo.rs"]
+    /// pub fn bar() {}
+    /// ```
+    ///
+    /// Use instead:
+    /// ```rust,ignore
+    /// // lib.rs
+    /// mod foo;
+    /// use crate::foo::bar;
+    /// ```
+    ///
+    /// ```rust,ignore
+    /// // foo.rs
+    /// #[path = "./foo.rs"]
+    /// pub fn bar() {}
+    /// ```
+    #[clippy::version = "1.84.0"]
+    pub USE_CRATE_PREFIX_FOR_SELF_IMPORTS,
+    nursery,
+    "checks that imports from the current crate use the `crate::` prefix"
+}
+
+#[derive(Clone, Default)]
+pub struct UseCratePrefixForSelfImports<'a, 'tcx> {
+    /// collect `use` in current block
+    use_block: Vec<&'a UsePath<'tcx>>,
+    /// collect `mod` in current block
+    mod_names: FxHashSet<Symbol>,
+    /// spans of `mod`, `use`, and attributes
+    spans: Vec<Span>,
+}
+
+impl_lint_pass!(UseCratePrefixForSelfImports<'_, '_> => [USE_CRATE_PREFIX_FOR_SELF_IMPORTS]);
+
+impl<'a, 'tcx> LateLintPass<'tcx> for UseCratePrefixForSelfImports<'a, 'tcx> {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'a Item<'tcx>) {
+        let FileName::Real(RealFileName::LocalPath(p)) = cx.sess().source_map().span_to_filename(item.span) else {
+            self.clear();
+            return;
+        };
+        let Some(file_name) = p.file_name() else {
+            self.clear();
+            return;
+        };
+        // only check `main.rs` and `lib.rs`
+        if !(file_name == "main.rs" || file_name == "lib.rs") {
+            return;
+        }
+
+        if self.in_same_block(item.span) {
+            self.insert_item(item);
+        } else {
+            self.deal(cx);
+            self.clear();
+            self.insert_item(item);
+        }
+    }
+}
+
+impl<'tcx> UseCratePrefixForSelfImports<'_, 'tcx> {
+    fn in_same_block(&self, span: Span) -> bool {
+        if self.spans.is_empty() {
+            return true;
+        }
+        if self.spans.iter().any(|x| x.contains(span)) {
+            return true;
+        }
+        if self.spans.iter().any(|x| span.lo() - x.hi() == BytePos(1)) {
+            return true;
+        }
+        false
```

Comment:
This can be shortened:

```suggestion
        self.spans.is_empty() || self.spans.iter().any(|x| x.contains(span) || span.lo() - x.hi() == BytePos(1))
```

What is the latest condition for? Isn't that equivalent to `span.shrink_to_lo() == span.shrink_to_hi()`? In which case would that trigger? This would deserve some documentation. The method should get documentatiojn too.

Reply:
Two adjacent lines
```
mod foo;
use foo;
```

Reply:
If `mod` and `use` in same block: don't lint

Reply:
Do you mean  `parent_hir_node`  or  `opt_parent` ?

Reply:
For example. Or maybe even `TypeCtxt::parent_module()` will do.

Reply:
I'm trying `check_attribute`, but blocked by https://github.com/rust-lang/rust/issues/146839